src/core_modules/capture-core/metaDataStoreLoaders/programs/loadPrograms.js

```diff
@@ -1,110 +1,171 @@
+    const apiProgramsObject = apiPrograms
+        .reduce((acc, apiProgram) => {
+            acc[apiProgram.id] = apiProgram;
+            return acc;
+        }, {});
+
+    const unavailableProgramIds = cachePrograms
+        .filter(cacheProgram => !apiProgramsObject[cacheProgram.id])
+        .map(unavailableProgram => unavailableProgram.id);
+
+    if (unavailableProgramIds.length > 0) {
+        const { storageController, storeNames } = getContext();
+        await storageController.remove(storeNames.PROGRAMS, unavailableProgramIds);
+    }
+};
+
+/**
+ * Retrieve the program ids for the programs we would like to update in the cache
+ * @param {Object[]} apiPrograms: programs from the api
+ * @param {Object[]} cachePrograms: programs from the cache
+ */
+const getStaleProgramIds = (apiPrograms, cachePrograms) => {
+    const cacheProgramsAsObject = cachePrograms
+        .reduce((acc, cacheProgram) => {
+            acc[cacheProgram.id] = cacheProgram;
+            return acc;
+        }, {});
+
+    return apiPrograms
+        .filter((program) => {
+            const cachedProgram = cacheProgramsAsObject[program.id];
+            return !cachedProgram || cachedProgram.version !== program.version;
+        })
+        .map(program => program.id);
+};
+
+/**
+ * Update the cache for the program ids passed in.
+ * @param {string[]} programIds: the program ids with stale data in the cache
+ * @returns programs data that is needed for the side effects.
+ */
+const loadProgramBatch = async (programIds) => {
+    const { convertedData: programs = [] } = await storePrograms(programIds);
+    await loadRulesCentricMetadata(programIds);
+    return programs
+        .map(program => ({
+            programTrackedEntityAttributes: program.programTrackedEntityAttributes,
+            categoryCombo: program.categoryCombo,
+            trackedEntityTypeId: program.trackedEntityTypeId,
+        }));
+};
+
+/**
+ * Self executing function to scope the side effect helper functions
+ */
+const getSideEffects = (() => {
+    const getOptionSetsOutline = (() => {
+        const getDataElementOptionSets = programStageDataElements =>
+            (programStageDataElements || [])
+                .map(psda => psda.dataElement && psda.dataElement.optionSet)
+                .filter(optionSet => optionSet);
+
+        const getTrackedEntityAttributeOptionSets = programTrackedEntityAttributes =>
+            (programTrackedEntityAttributes || [])
+                .map(ptea => ptea.trackedEntityAttribute && ptea.trackedEntityAttribute.optionSet)
+                .filter(optionSet => optionSet);
+
+        const getProgramOptionSets = (program) => {
+            const dataElementOptionSets = (program.programStages || [])
+                .flatMap(programStage => getDataElementOptionSets(programStage.programStageDataElements));
+
+            return [
+                ...dataElementOptionSets,
```

Comment:
Sorry I misspelled. My point was more on the fact that `getProgramOptionSets` and the 
```
(program.programStages || [])
                .flatMap(programStage => getDataElementOptionSets(programStage.programStageDataElements))
```

could be merged into one function that what would do is return the `dataElementOptionSets`. Now I dont understand what is the point of having the 
```
(program.programStages || [])
                .flatMap(programStage => getDataElementOptionSets(programStage.programStageDataElements))
```

as a separate calculation

Reply:
Sorry for not explaining my self again very well. 
I dont mean to say that the code is unreasonable only thing I am trying to say is that we might wanna improve the readability so that you as the author is not the only one who understands the code but also the rest of the people who might read it.

I didnt meant to get into such a long discussion about this but my suggestion sits between these lines 

```
const getOptionSetsOutline = (() => {
        const getDataElementOptionSets = programStages => returns the data element option sets

        const getTrackedEntityAttributeOptionSets = programTrackedEntityAttributes => returns the tracked entity attribute options sets

        return (programsOutline): Array<Object> =>
            programsOutline
                .flatMap(program => [
                    ...getDataElementOptionSets(program.programStages),
                    ...getTrackedEntityAttributeOptionSets(program.programTrackedEntityAttributes),
                ]);
    })();
```

Reply:
I disagree (obviously). But since this discussion over a minor suggestion has been going on for far too long (and I really need this PR merged), I'm changing it.